lasagne/layers/pool.py

```diff
@@ -15,81 +14,212 @@
+    ignore_border: bool
+        If True, partial pooling regions will be ignored.
 
     Returns
     -------
     output_length
-        * None if either input is None
-        * `input_length / pool_length` otherwise
-    '''
-
-    if input_length is None or pool_length is None:
+        * None if either input is None.
+        * Computed length of the pooling operator otherwise.
+
+    Notes
+    -----
+    When ignore_border = True, this is given by the number of full
+    pooling regions that fit in the padded input length,
+    divided by the stride (rounding down).
+
+    When ignore_border = False, partial pooling regions are also
+    allowed, as long as each pooling region contains at least one element
+    not present in another pooling region.
+    """
+    if input_length is None or ds is None:
         return None
 
     if ignore_border:
-        # Round shape down when we ignore the border
-        return input_length // pool_length
+        output_length = input_length + 2 * pad - ds + 1
+        output_length = (output_length + st - 1) // st
+
+    # output length calculation taken from:
+    # https://github.com/Theano/Theano/blob/master/theano/tensor/signal/downsample.py
+    else:
+        if st >= ds:
+            output_length = (input_length + st - 1) // st
+        else:
+            output_length = max(
+                0, (input_length - ds + st - 1) // st) + 1
 
-    # Otherwise, round the shape up
-    return int(np.ceil(float(input_length) / pool_length))
+    return output_length
 
 
 class MaxPool1DLayer(Layer):
-    def __init__(self, incoming, ds, ignore_border=False, **kwargs):
+    """
+    This layer performs max pooling over the final dimension
+    of a 3D tensor.
+    """
+    def __init__(self, incoming, ds, st=None, pad=0, ignore_border=False,
+                 **kwargs):
+        """
+        Instantiates the layer.
+
+        Parameters
+        ----------
+        incoming : a :class:`Layer` instance or tuple
+            The layer feeding into this layer, or the expected input shape.
+        ds : integer
+            The length of the pooling region
+        st : integer or None
+            The stride between sucessive pooling regions.
+            If None, st = ds.
+        pad : integer
+            The number of elements to be added to the input on each side.
+            Must be less than st.
+        ignore_border : bool
+            If True, partial pooling regions will be ignored.
+            Must be True if pad != 0.
+
+        Notes
+        -----
+        The value used to pad the input is chosen to be less than
+        the minimum of the input, so that the output of each pooling region
+        always corresponds to some element in the unpadded input region.
+        """
         super(MaxPool1DLayer, self).__init__(incoming, **kwargs)
         self.ds = ds  # an integer
+        self.st = ds if st is None else st
+        self.pad = pad
         self.ignore_border = ignore_border
```

Comment:
I thought that the `ignore_border`/`pad` incompatibility needs to be checked somewhere, but then I had a look at the Theano implementation to see that [it raises a `NotImplementedError`](https://github.com/Theano/Theano/blob/master/theano/tensor/signal/downsample.py#L214-216). I think it's good to leave it up to Theano along with all other input checking then.